codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -886,12 +886,14 @@ public void close(ChannelPromise promise) {
                         }
                     }
 
+                    // The promise should be notified before we call fireChannelInactive().
+                    promise.setSuccess();
```

Comment:
I also think the order of these two needs to be reversed, the following test fails:
```java
@Test
    public void channelClosedWhenClosePromiseCompletes() {
        LastInboundHandler inboundHandler = streamActiveAndWriteHeaders(inboundStream);
        Http2StreamChannel childChannel = (Http2StreamChannel) inboundHandler.channel();

        assertTrue(childChannel.isOpen());
        assertTrue(childChannel.isActive());

        final AtomicBoolean channelOpen = new AtomicBoolean(true);

        final ChannelPromise p = childChannel.newPromise();

        p.addListener(new ChannelFutureListener() {
            @Override
            public void operationComplete(ChannelFuture future) throws Exception {
                channelOpen.set(future.channel().isOpen());
            }
        });

        childChannel.close(p).syncUninterruptibly();

        assertFalse(channelOpen.get());
    }
```

Reversing the order fixes the test above. I'm happy to file another bug if you prefer (perhaps that breaks something else I'm not aware of).

Reply:
+1 ... there was actually another issue too so I just submitted https://github.com/netty/netty/pull/7643 with a modified version of the unit test above.